src/detect.rs

```diff
@@ -18,3 +23,196 @@ pub enum DetectOutcome {
+}
+
+enum PlatformArg<P> {
+    Plat(P),
+    Dir(PathBuf),
+}
+
+pub struct DetectContextBuilder<P: Platform, BM: DeserializeOwned> {
+    app_dir: Option<PathBuf>,
+    buildpack_dir: Option<PathBuf>,
+    stack_id: String,
+    platform: Option<PlatformArg<P>>,
+    buildpack_toml: Option<BuildpackTomlArg<BM>>,
+}
+impl<P: Platform, BM: DeserializeOwned> DetectContextBuilder<P, BM> {
+    pub fn new(stack_id: impl AsRef<str>) -> Self {
+        DetectContextBuilder {
+            app_dir: None,
+            buildpack_dir: None,
+            stack_id: String::from(stack_id.as_ref()),
+            platform: None,
+            buildpack_toml: None,
+        }
+    }
+
+    pub fn buildpack_toml(mut self, arg: BuildpackToml<BM>) -> DetectContextBuilder<P, BM> {
+        self.buildpack_toml = Some(BuildpackTomlArg::BuildpackToml(arg));
+        self
+    }
+
+    pub fn buildpack_toml_from_string(
+        mut self,
+        string: impl AsRef<str>,
+    ) -> DetectContextBuilder<P, BM> {
+        self.buildpack_toml = Some(BuildpackTomlArg::Str(string.as_ref().to_string()));
+        self
+    }
+
+    pub fn buildpack_toml_from_path(
+        mut self,
+        path: impl AsRef<Path>,
+    ) -> DetectContextBuilder<P, BM> {
+        self.buildpack_toml = Some(BuildpackTomlArg::Path(path.as_ref().to_path_buf()));
+        self
+    }
+
+    /// Used for testing purposes to set a minimal viable context
+    ///
+    /// Warning: The directory passed in will be modified.
+    ///
+    ///```
+    /// use tempfile;
+    /// use libcnb::DetectContextBuilder;
+    ///
+    /// pub fn detect(context: libcnb::GenericDetectContext) {}
+    ///
+    /// let temp_dir = tempfile::tempdir().unwrap();
+    /// let context = DetectContextBuilder::new("heroku-20")
+    ///     .buildpack_toml_from_string(include_str!("../examples/example-02-ruby-sample/buildpack.toml"))
+    ///     .from_path(&temp_dir)
+    ///     .build::<std::io::Error>()
+    ///     .unwrap();
+    ///
+    /// detect(context);
+    ///```
+    pub fn from_path(mut self, dir: impl AsRef<Path>) -> DetectContextBuilder<P, BM> {
+        let dir = dir.as_ref();
+        let app_dir = dir.join("app");
+        let platform_dir = dir.join("platform");
+        let buildpack_dir = dir.join("buildpack");
+
+        for dir in [&app_dir, &platform_dir, &buildpack_dir] {
+            std::fs::create_dir_all(dir).unwrap();
+        }
+
+        if self.app_dir.is_none() {
+            self.app_dir = Some(app_dir);
+        }
+
+        if self.buildpack_dir.is_none() {
+            self.buildpack_dir = Some(buildpack_dir);
+        }
+
+        if self.platform.is_none() {
+            self.platform = Some(PlatformArg::Dir(platform_dir));
+        }
+        self
+    }
+
+    pub fn platform(mut self, platform: P) -> DetectContextBuilder<P, BM> {
+        self.platform = Some(PlatformArg::Plat(platform));
+        self
+    }
+
+    pub fn platform_from_path(
```

Comment:
Naming: Avoiding the `from` prefix, this could be called `platform_path`. For consistency, the naming for `app_dir` is slightly different too.

Reply:
Resolved by moving all this logic to a single `platform` function.